### ISSUE TITLE
Blueprints are conceptually coupled to apps before registration

### ISSUE THREAD

**Issue by nicbon** [jfinkels/flask-restless#397]
I'm testing the new 0.16 release.
I get a

    AttributeError: 'NoneType' object has no attribute 'extensions'
after calling create_api_blueprint. I think this happens because I don't have initialized the APIManager with an app instance nor I specify my app instance in the create_api_blueprint method.

Having to specify the app at the time of blueprint creation couples apps and blueprints together before the blueprints are registered.

Ideally, it should not be enforced to know the app instance to which the blueprint will be attached before it is registered to it. I don't master the details of Flask-Restless internals but I noticed that the APIManager uses the app instance to know the Flask-Restless configuration details via

    app.extensions['restless']
and the current registered blueprints

   app.blueprints.

For app.extensions, I think that the configuration details such as global pre- and post- processors can be retrieved from the APIManager instance itself. Each blueprint created by a specific APIManager instance will inherit its global configuration. Multiple instances of APIManager can provide different configuration settings.

For app.blueprints, I noticed that it is used in the _next_blueprint_name. If the point of _next_blueprint_name is to generate unique names, I think the app.blueprints information is not needed. If you want to use the current name semantic it should be possible to derive a new class from Blueprint and override the register method as suggested in the Flask docs.

Removing these two references will keep the decoupling between apps and blueprints intact.

What do you think?

**Comment by nicbon** [comment]
I'm diving more into the code the of APIManager. To me, It seems more streamlined to require that the restless info are specified at construction time, i.e.
```python
def __init__(self, app=None, session=None, flask_sqlalchemy_db=None,
    preprocessors=None, postprocessors=None):
```
These are parameters that characterize your APIManager instance whatever application it will be registered to. Sanity checks can be performed in the constructor.
```python
if session is None:
    try:
        session = flask_sqlalchemy_db.sessions
    except AttributeError:
        raise ValueError('Flask-Restless requires a valid flask_sqlalchemy_db or session')
```
The RestlessInfo object can be stored in the manager itself (it can then be passed to the app at init time) although passing the whole APIManager could be another option.

Also, I rather require name (string) instead of app as argument for create_api_blueprint, giving the users the possibility to name their blueprints and provide a unique identifier (e.g. uuid1?) when blueprints are created automatically.

I'm working on a fork that implements these changes.

**Comment by arthurbarros** [comment]
+1

Is it will be merged?

**Comment by jfinkels** [comment]
This should be fixed due to #404 and b6c92623964e7aea899a20f17ee66f1f1bd5cc39. Please test the development version and reopen a new issue if you are still having this problem.
